File: app/src/main/java/com/amaze/filemanager/ui/notifications/FTPNotification.java
package com.amaze.filemanager.ui.notifications;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.preference.PreferenceManager;
import android.support.v4.app.NotificationCompat;

import com.amaze.filemanager.R;
import com.amaze.filemanager.activities.MainActivity;
import com.amaze.filemanager.asynchronous.services.ftp.FTPService;

import java.net.InetAddress;

/**
 * Created by yashwanthreddyg on 19-06-2016.
 */
public class FTPNotification extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        switch(intent.getAction()){
            case FTPService.ACTION_STARTED:
                createNotification(context, intent.getBooleanExtra(FTPService.TAG_STARTED_BY_TILE, false));
                break;
            case FTPService.ACTION_STOPPED:
                removeNotification(context);
                break;
        }
    }

    private void createNotification(Context context, boolean noStopButton) {

        String notificationService = Context.NOTIFICATION_SERVICE;
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(notificationService);

        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        int port = sharedPreferences.getInt(FTPService.PORT_PREFERENCE_KEY, FTPService.DEFAULT_PORT);
        boolean secureConnection = sharedPreferences.getBoolean(FTPService.KEY_PREFERENCE_SECURE, FTPService.DEFAULT_SECURE);

        InetAddress address = FTPService.getLocalInetAddress(context);

        String iptext = (secureConnection ? FTPService.INITIALS_HOST_SFTP : FTPService.INITIALS_HOST_FTP)
                + address.getHostAddress() + ":"
                + port + "/";

        int icon = R.drawable.ic_ftp_light;
        CharSequence tickerText = context.getResources().getString(R.string.ftp_notif_starting);
        long when = System.currentTimeMillis();


        CharSequence contentTitle = context.getResources().getString(R.string.ftp_notif_title);
        CharSequence contentText = String.format(context.getResources().getString(R.string.ftp_notif_text), iptext);

        Intent notificationIntent = new Intent(context, MainActivity.class);
        notificationIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        PendingIntent contentIntent = PendingIntent.getActivity(context, 0, notificationIntent, 0);

        NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(context, NotificationConstants.CHANNEL_FTP_ID)
                .setContentTitle(contentTitle)
                .setContentText(contentText)
                .setContentIntent(contentIntent)
                .setSmallIcon(icon)
                .setTicker(tickerText)
                .setWhen(when)
                .setOngoing(true);

        NotificationConstants.setMetadata(context, notificationBuilder, NotificationConstants.TYPE_FTP);

        Notification notification;
        if (!noStopButton && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            int stopIcon = android.R.drawable.ic_menu_close_clear_cancel;
            CharSequence stopText = context.getResources().getString(R.string.ftp_notif_stop_server);
            Intent stopIntent = new Intent(FTPService.ACTION_STOP_FTPSERVER);
            PendingIntent stopPendingIntent = PendingIntent.getBroadcast(context, 0,
                    stopIntent, PendingIntent.FLAG_ONE_SHOT);

            notificationBuilder.addAction(stopIcon, stopText, stopPendingIntent);
            notificationBuilder.setShowWhen(false);
            notification = notificationBuilder.build();
        } else {
            notification = notificationBuilder.getNotification();
        }

        // Pass Notification to NotificationManager
        notificationManager.notify(NotificationConstants.FTP_ID, notification);
    }

    private void removeNotification(Context context){
        String ns = Context.NOTIFICATION_SERVICE;
        NotificationManager nm = (NotificationManager) context.getSystemService(ns);
        nm.cancelAll();
    }
}

File: app/src/main/java/com/amaze/filemanager/ui/notifications/FtpNotification.kt
/*
 * Copyright (C) 2014-2020 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
 * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
 *
 * This file is part of Amaze File Manager.
 *
 * Amaze File Manager is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.amaze.filemanager.ui.notifications

import android.annotation.SuppressLint
import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import androidx.annotation.StringRes
import androidx.core.app.NotificationCompat
import androidx.preference.PreferenceManager
import com.amaze.filemanager.R
import com.amaze.filemanager.asynchronous.services.ftp.FtpService
import com.amaze.filemanager.asynchronous.services.ftp.FtpService.Companion.getLocalInetAddress
import com.amaze.filemanager.ui.activities.MainActivity
import com.amaze.filemanager.ui.notifications.NotificationConstants.setMetadata

/**
 * Created by yashwanthreddyg on 19-06-2016.
 * Edited by zent-co on 30-07-2019.
 * Re-edited by hojat72elect on 14-05-2022.
 */
object FtpNotification {
    @SuppressLint("UnspecifiedImmutableFlag")
    private fun buildNotification(
        context: Context,
        @StringRes contentTitleRes: Int,
        contentText: String,
        noStopButton: Boolean
    ): NotificationCompat.Builder {
        val notificationIntent = Intent(context, MainActivity::class.java)
        notificationIntent.flags =
            Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
        val contentIntent = PendingIntent.getActivity(context, 0, notificationIntent, 0)
        val currentTime = System.currentTimeMillis()
        val builder = NotificationCompat.Builder(context, NotificationConstants.CHANNEL_FTP_ID)
            .setContentTitle(context.getString(contentTitleRes))
            .setContentText(contentText)
            .setContentIntent(contentIntent)
            .setSmallIcon(R.drawable.ic_ftp_light)
            .setTicker(context.getString(R.string.ftp_notif_starting))
            .setWhen(currentTime)
            .setOngoing(true)
            .setOnlyAlertOnce(true)
        if (!noStopButton) {
            val stopIcon = android.R.drawable.ic_menu_close_clear_cancel
            val stopText: CharSequence = context.getString(R.string.ftp_notif_stop_server)
            val stopIntent =
                Intent(FtpService.ACTION_STOP_FTPSERVER).setPackage(context.packageName)
            val stopPendingIntent =
                PendingIntent.getBroadcast(context, 0, stopIntent, PendingIntent.FLAG_ONE_SHOT)
            builder.addAction(stopIcon, stopText, stopPendingIntent)
        }
        setMetadata(context, builder, NotificationConstants.TYPE_FTP)
        return builder
    }

    fun startNotification(context: Context, noStopButton: Boolean): Notification {
        val builder = buildNotification(
            context,
            R.string.ftp_notif_starting_title,
            context.getString(R.string.ftp_notif_starting),
            noStopButton
        )
        return builder.build()
    }

    fun updateNotification(context: Context, noStopButton: Boolean) {
        val notificationService = Context.NOTIFICATION_SERVICE
        val notificationManager =
            context.getSystemService(notificationService) as NotificationManager
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        val port = sharedPreferences.getInt(FtpService.PORT_PREFERENCE_KEY, FtpService.DEFAULT_PORT)
        val secureConnection = sharedPreferences.getBoolean(
            FtpService.KEY_PREFERENCE_SECURE,
            FtpService.DEFAULT_SECURE
        )
        val address = getLocalInetAddress(context)
        var address_text = "Address not found"
        if (address != null) {
            address_text =
                (
                    (
                        if (secureConnection) FtpService.INITIALS_HOST_SFTP
                        else FtpService.INITIALS_HOST_FTP
                        ) +
                        address.hostAddress +
                        ":" +
                        port +
                        "/"
                    )
        }
        val builder = buildNotification(
            context,
            R.string.ftp_notif_title,
            context.getString(R.string.ftp_notif_text, address_text),
            noStopButton
        )
        notificationManager.notify(NotificationConstants.FTP_ID, builder.build())
    }

    private fun removeNotification(context: Context) {
        val ns = Context.NOTIFICATION_SERVICE
        val nm = context.getSystemService(ns) as NotificationManager
        nm.cancelAll()
    }
}
